src/evalfunc.c

```diff
@@ -2446,12 +2446,30 @@ f_char2nr(typval_T *argvars, typval_T *rettv)
     if (has_mbyte)
     {
 	int	utf8 = 0;
+	int	retlist = 0;
 
 	if (argvars[1].v_type != VAR_UNKNOWN)
+	{
 	    utf8 = (int)tv_get_number_chk(&argvars[1], NULL);
+	    if (argvars[2].v_type != VAR_UNKNOWN)
+		retlist = (int)tv_get_number_chk(&argvars[2], NULL);
+	}
 
-	if (utf8)
-	    rettv->vval.v_number = (*utf_ptr2char)(tv_get_string(&argvars[0]));
+	if (utf8 || enc_utf8)
+	{
+	    char_u *p = tv_get_string(&argvars[0]);
+
+	    if (retlist)
+	    {
+		if (rettv_list_alloc(rettv) == FAIL)
+		    return;
+		for (; *p != NUL; p += utf_ptr2len(p))
+		    list_append_number(rettv->vval.v_list,
+						(varnumber_T)utf_ptr2char(p));
```

Comment:
With this, the returned value will include *all* characters of the given string - not only the first one plus composing characters. If this is really what we wanit, I think it would be nice to warn about this more clearly in the documentation.